components/encryption/src/log_file.rs

```diff
@@ -0,0 +1,326 @@
+        file_dict.merge_from_bytes(content)?;
+
+        // parse the remained records
+        while !remained.is_empty() {
+            let (used_size, file_name, mode) = parse_next_record(remained)?;
+            remained.consume(used_size);
+            match mode {
+                LogRecord::INSERT(info) => {
+                    file_dict.files.insert(file_name, info);
+                }
+                LogRecord::REMOVE => {
+                    file_dict.files.remove(&file_name);
+                }
+            }
+        }
+
+        Ok(file_dict)
+    }
+
+    /// Append an insert operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn insert(&self, name: String, info: FileInfo) -> Result<()> {
+        let mut file = self.file_lock.lock().unwrap();
+        if let Some(file) = file.as_mut() {
+            let bytes = convert_record_to_bytes(&name, LogRecord::INSERT(info))?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+
+    /// Append a remove operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn remove(&self, name: &str) -> Result<()> {
+        let mut file = self.file_lock.lock().unwrap();
+        if let Some(file) = file.as_mut() {
+            let bytes = convert_record_to_bytes(&name, LogRecord::REMOVE)?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+}
+
+/// Header of record.
+///
+/// ```text
+/// +----+--+--+-+----+----
+/// |    |  |  | |    |
+/// ++---++-++-++++---++---
+///  ^    ^  ^  ^ ^    ^
+///  |    |  |  | |    + FileInfo content (variable bytes)
+///  |    |  |  | + File name content (variable bytes)
+///  |    |  |  + Log type (1 bytes)
+///  |    |  + FileInfo length (2 bytes)
+///  |    + File name length (2 bytes)
+///  + Crc32 (4 bytes)
+/// ```
+const RECORD_HEADER_SIZE: usize = 4 + 2 + 2 + 1;
+
+fn convert_record_to_bytes(fname: &str, record_type: LogRecord) -> Result<Vec<u8>> {
+    let name_len = fname.len() as u16;
+    let mut content = vec![];
+    content.extend_from_slice(fname.as_bytes());
+
+    let mut header_buf = [0; RECORD_HEADER_SIZE];
+    let info_len = match record_type {
+        LogRecord::INSERT(info) => {
+            header_buf[8] = 1;
+            let info_bytes = info.write_to_bytes()?;
+            content.extend_from_slice(&info_bytes);
+            info_bytes.len() as u16
+        }
+        LogRecord::REMOVE => {
+            header_buf[8] = 2;
+            0
+        }
+    };
+
+    let mut digest = crc32fast::Hasher::new();
+    digest.update(&content);
```

Comment:
crc should include all the rest of record (filename len, file info len, and log type) as well.

Reply:
Refer to `encrypted_file::header`, crc only checks the content part. We need a length to start the check. If the length is unknown, the check range cannot be obtained.